wcwidth/wcwidth.py

```diff
@@ -154,35 +127,10 @@ def wcwidth(wc):
           Full-width (F) category as defined in Unicode Technical
           Report #11 have a column width of 2.
     """
-    # pylint: disable=C0103
-    #         Invalid argument name "wc"
-    ucs = ord(wc)
-
-    # NOTE: created by hand, there isn't anything identifiable other than
-    # general Cf category code to identify these, and some characters in Cf
-    # category code are of non-zero width.
-    if (0 == ucs or
-            0x034F == ucs or
-            0x200B <= ucs <= 0x200F or
-            0x2028 == ucs or
-            0x2029 == ucs or
-            0x202A <= ucs <= 0x202E or
-            0x2060 <= ucs <= 0x2063):
-        return 0
-
-    # C0/C1 control characters
-    if ucs < 32 or 0x07F <= ucs < 0x0A0:
-        return -1
-
-    # combining characters have indeterminate effects unless
-    # combined with additional characters.
-    if _bisearch(ucs, NONZERO_COMBINING):
-        return -1
-
-    return 1 + _bisearch(ucs, WIDE_EASTASIAN)
-
-
-def wcswidth(pwcs, n=None):
+    return wcswidth(wc, negative_returnval=True)
+
+
+def wcswidth(pwcs, n=None, negative_returnval=False):
```

Comment:
Yes. This is kind of a hack.

In the original code, wcswidth would repeatedly call wcwidth, leading to a high function call overhead. Having wcwidth call wcswidth with a one-char string instead thus improves performance. The problem is, that the original wcwidth would return -1 for control characters or combining characters, where wcswidth would return -1 if it found a control character and ignore any combining characters (treat them as width 0).

I'll think about this some more. I don't really like that additional argument, and at the very least it needs a more descriptive name. Maybe it would be ok to just duplicate the control/combining character detection in wcwidth.